libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,267 @@
+	// Sleep, read again. If we still don't have anything, assume the other PID is dead, continue.
+	if !gotData {
+		lock.logger.Debugf("Found other pidfile, but no data. Retrying.")
+		time.Sleep(time.Millisecond * 500)
+		gotData, pf, err = lock.readExistingPidfile()
+		if err != nil {
+			return fmt.Errorf("error re-reading existing lockfile: %w", err)
+		}
+		if !gotData {
+			lock.logger.Debugf("No PID found in other lockfile, continuing")
+			return lock.recoverLockfile()
+		}
+	}
+
+	// Case: the lockfile is locked, but by us. Probably a coding error,
+	// and probably hard to do
+	if pf.Pid == os.Getpid() {
+		// the lockfile was written before the beat started, meaning we restarted and somehow got the same pid
+		// in which case, continue
+		if lock.beatStart.Before(pf.WriteTime) {
+			return fmt.Errorf("lockfile for beat has been locked twice by the same PID, potential bug.")
+		}
+		lock.logger.Debugf("Beat has started with the same PID, continuing")
+		return lock.recoverLockfile()
+	}
+
+	// Check to see if the PID found in the pidfile exists.
+	existsState, err := findMatchingPID(pf.Pid)
+	// Case: we have a lockfile, but the pid from the pidfile no longer exists
+	// this was presumably due to the dirty shutdown.
+	// Try to reset the lockfile and continue.
+	if errors.Is(err, metricproc.ProcNotExist) {
+		lock.logger.Debugf("%s shut down without removing previous lockfile, continuing", lock.beatName)
+		return lock.recoverLockfile()
+	} else if err != nil {
+		return fmt.Errorf("error looking up status for pid %d: %w", pf.Pid, err)
+	} else {
+		// Case: the PID exists, but it's attached to a zombie process
+		// In this case...we should be okay to restart?
+		if existsState == metricproc.Zombie {
+			lock.logger.Debugf("%s shut down without removing previous lockfile and is currently in a zombie state, continuing", lock.beatName)
+			return lock.recoverLockfile()
+		}
+		// Case: we've gotten a lock file for another process that's already running
+		// This is the "base" lockfile case, which is two beats running from the same directory
+		// This is where we'll catch this particular case on Linux, due to Linux's advisory-style locks.
+		return fmt.Errorf("connot start, data directory belongs to process with pid %d", pf.Pid)
+	}
+}
+
+// recoverLockfile attempts to remove the lockfile and continue running
+// This should only be called after we're sure it's safe to ignore a pre-existing lockfile
+// This will reset the internal lockfile path when it's successful.
+func (lock *Locker) recoverLockfile() error {
+	// File remove may or not work, depending on os-specific details with lockfiles
+	err := os.Remove(lock.fileLock.Path())
+	if err != nil {
+		lock.logger.Debugf("Could not remove old lockfile, got error %s. Attempting to fix.")
+		rname := rand.New(rand.NewSource(time.Now().UnixNano()))
+		lockfilePath := paths.Resolve(paths.Data, fmt.Sprintf("%s_%d.lock", lock.beatName, rname.Int()))
```

Comment:
I think this might be a problem.  What happens in following scenario.

1) beat instance 1 dies, leaves lock file, and for some reason os.Remove will fail on it
2) beat instance 2 starts, can't remove lock file, makes new one with random number in name
3) beat instance 3 starts, can't remove lock file, makes a new one with different random number in name

I think at that point we have beat instance 2 & 3 running, with different lock files but in the same directory.  Is that what would happen?

If so, I'm thinking if `os.Remove` fails the code should fail and let a human figure it out.

Reply:
Huh, I didn't even think of that.

Yah, now that I think of it, you might be right. The two main cases we're trying to cover with this PR is "lockfile isn't removed automatically, can be removed at next startup" and "must wait for OS to release lock on windows."

Neither of those cases really fall under the issue of "file can not be removed at all" which is what this covers.

Thinking I'll remove this, and it should simplify things a good bit.